File: collect_app/src/test/java/org/odk/collect/android/activities/FormMapActivityTest.java
package org.odk.collect.android.activities;

import static org.hamcrest.CoreMatchers.is;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;
import static org.odk.collect.android.activities.viewmodels.FormMapViewModel.ClickAction.DELETED_TOAST;
import static org.odk.collect.android.activities.viewmodels.FormMapViewModel.ClickAction.NOT_VIEWABLE_TOAST;
import static org.odk.collect.android.activities.viewmodels.FormMapViewModel.ClickAction.OPEN_EDIT;
import static org.odk.collect.android.activities.viewmodels.FormMapViewModel.ClickAction.OPEN_READ_ONLY;

import android.content.Intent;
import android.util.Pair;
import android.view.View;
import android.widget.TextView;

import androidx.test.ext.junit.runners.AndroidJUnit4;

import com.google.android.material.chip.Chip;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.ArgumentMatchers;
import org.odk.collect.android.R;
import org.odk.collect.android.activities.viewmodels.FormMapViewModel;
import org.odk.collect.android.external.InstanceProvider;
import org.odk.collect.android.geo.MapProvider;
import org.odk.collect.android.geo.TestMapFragment;
import org.odk.collect.android.injection.config.AppDependencyComponent;
import org.odk.collect.android.injection.config.AppDependencyModule;
import org.odk.collect.android.support.CollectHelpers;
import org.odk.collect.forms.Form;
import org.odk.collect.forms.FormsRepository;
import org.odk.collect.forms.instances.Instance;
import org.odk.collect.forms.instances.InstancesRepository;
import org.odk.collect.formstest.FormUtils;
import org.odk.collect.geo.maps.MapPoint;
import org.odk.collect.shared.TempFiles;
import org.robolectric.android.controller.ActivityController;

import java.util.Arrays;
import java.util.Date;
import java.util.List;

@RunWith(AndroidJUnit4.class)
public class FormMapActivityTest {
    private FormMapActivity activity;

    private final TestMapFragment map = new TestMapFragment();

    private final List<MapPoint> expectedPoints = Arrays.asList(new MapPoint(10.0, 125.6),
            new MapPoint(10.1, 125.6), new MapPoint(10.1, 126.6),
            new MapPoint(10.3, 125.6), new MapPoint(10.3, 125.7),
            new MapPoint(10.4, 125.6));

    @Before
    public void setUpActivity() {
        CollectHelpers.setupDemoProject();

        AppDependencyComponent component = CollectHelpers.overrideAppDependencyModule(new AppDependencyModule() {
            @Override
            public MapProvider providesMapProvider() {
                MapProvider mapProvider = mock(MapProvider.class);
                when(mapProvider.createMapFragment(ArgumentMatchers.any())).thenReturn(map);
                return mapProvider;
            }
        });

        FormsRepository formsRepository = component.formsRepositoryProvider().get();
        InstancesRepository instancesRepository = component.instancesRepositoryProvider().get();
        Form form = formsRepository.save(testForm);
        Arrays.stream(testInstances).forEach(instancesRepository::save);

        Intent intent = new Intent();
        intent.putExtra(FormMapActivity.EXTRA_FORM_ID, form.getDbId());
        ActivityController activityController = CollectHelpers.buildThemedActivity(FormMapActivity.class, intent);
        activity = (FormMapActivity) activityController.get();
        activityController.setup();
    }

    // Note that there's a point with deleted status included. This shouldn't be possible in real
    // usage because deleting a form removes the geometry from the database. However, the database
    // allows a deleted instance with geometry so we test it.
    @Test
    public void mappedPoints_matchInstancesWithGeometry() {
        assertThat(map.getMappedPointCount(), is(expectedPoints.size()));
        for (MapPoint expectedPoint : expectedPoints) {
            assertThat(map.isMapped(expectedPoint), is(true));
        }
    }

    @Test
    public void tappingOnEditableInstance_showsSubmissionSummaryWithAppropriateMessage() {
        Pair<Instance, MapPoint> editableAndFinalized = new Pair<>(testInstances[1], new MapPoint(10.1, 125.6));
        Pair<Instance, MapPoint> unfinalized = new Pair<>(testInstances[2], new MapPoint(10.1, 126.6));
        Pair<Instance, MapPoint> failedToSend = new Pair<>(testInstances[4], new MapPoint(10.3, 125.6));

        activity.onFeatureClicked(map.getFeatureIdFor(editableAndFinalized.second));
        assertSubmissionSummaryContent(editableAndFinalized.first.getDisplayName(), editableAndFinalized.first.getStatus(), new Date(editableAndFinalized.first.getLastStatusChangeDate()), OPEN_EDIT);

        activity.onFeatureClicked(map.getFeatureIdFor(unfinalized.second));
        assertSubmissionSummaryContent(unfinalized.first.getDisplayName(), unfinalized.first.getStatus(), new Date(unfinalized.first.getLastStatusChangeDate()), OPEN_EDIT);

        activity.onFeatureClicked(map.getFeatureIdFor(failedToSend.second));
        assertSubmissionSummaryContent(failedToSend.first.getDisplayName(), failedToSend.first.getStatus(), new Date(failedToSend.first.getLastStatusChangeDate()), OPEN_READ_ONLY);
    }

    @Test
    public void tappingOnUneditableInstances_showsSubmissionSummaryWithAppropriateMessage() {
        Pair<Instance, MapPoint> sent = new Pair<>(testInstances[5], new MapPoint(10.3, 125.7));

        int featureId = map.getFeatureIdFor(sent.second);
        activity.onFeatureClicked(featureId);
        assertSubmissionSummaryContent(sent.first.getDisplayName(), sent.first.getStatus(), new Date(sent.first.getLastStatusChangeDate()), OPEN_READ_ONLY);
    }

    @Test
    public void tappingOnFailedSubmission_showsSubmissionSummaryWithAppropriateMessage() {
        Pair<Instance, MapPoint> submissionFailedCantEditWhenFinalized = new Pair<>(testInstances[6], new MapPoint(10.4, 125.6));

        int featureId = map.getFeatureIdFor(submissionFailedCantEditWhenFinalized.second);
        activity.onFeatureClicked(featureId);
        assertSubmissionSummaryContent(submissionFailedCantEditWhenFinalized.first.getDisplayName(), submissionFailedCantEditWhenFinalized.first.getStatus(), new Date(submissionFailedCantEditWhenFinalized.first.getLastStatusChangeDate()), NOT_VIEWABLE_TOAST);
    }

    // Geometry is removed from the database on instance deletion but just in case there is a
    // deleted instance with geometry available, show a deleted toast.
    @Test
    public void tappingOnDeletedInstances_showsSubmissionSummaryWithAppropriateMessage() {
        Pair<Instance, MapPoint> deleted = new Pair<>(testInstances[0], new MapPoint(10.0, 125.6));

        int featureId = map.getFeatureIdFor(deleted.second);
        activity.onFeatureClicked(featureId);
        assertSubmissionSummaryContent(deleted.first.getDisplayName(), deleted.first.getStatus(), new Date(deleted.first.getLastStatusChangeDate()), DELETED_TOAST);
    }

    private void assertSubmissionSummaryContent(String instanceName, String status, Date lastStatusChangeDate, FormMapViewModel.ClickAction clickAction) {
        assertThat(((TextView) activity.findViewById(R.id.name)).getText().toString(), is(instanceName));
        String instanceLastStatusChangeDate = InstanceProvider.getDisplaySubtext(activity, status, lastStatusChangeDate);
        assertThat(((TextView) activity.findViewById(R.id.status_text)).getText().toString(), is(instanceLastStatusChangeDate));

        switch (clickAction) {
            case DELETED_TOAST:
                assertThat(activity.findViewById(R.id.info).getVisibility(), is(View.VISIBLE));
                assertThat(activity.findViewById(R.id.action).getVisibility(), is(View.GONE));
                break;
            case NOT_VIEWABLE_TOAST:
                assertThat(activity.findViewById(R.id.info).getVisibility(), is(View.VISIBLE));
                assertThat(((TextView) activity.findViewById(R.id.info)).getText().toString(), is(activity.getString(R.string.cannot_edit_completed_form)));
                assertThat(activity.findViewById(R.id.action).getVisibility(), is(View.GONE));
                break;
            case OPEN_READ_ONLY:
                assertThat(activity.findViewById(R.id.info).getVisibility(), is(View.GONE));
                assertThat(activity.findViewById(R.id.action).getVisibility(), is(View.VISIBLE));
                assertThat(((Chip) activity.findViewById(R.id.action)).getText(), is(activity.getString(R.string.view_data)));
                break;
            case OPEN_EDIT:
                assertThat(activity.findViewById(R.id.info).getVisibility(), is(View.GONE));
                assertThat(activity.findViewById(R.id.action).getVisibility(), is(View.VISIBLE));
                assertThat(((Chip) activity.findViewById(R.id.action)).getText(), is(activity.getString(R.string.review_data)));
                break;
        }
    }

    private final Form testForm = FormUtils
            .buildForm("formId1", "2019103101", TempFiles.createTempDir().getAbsolutePath())
            .build();

    private static Instance[] testInstances = {
            new Instance.Builder()
                    .instanceFilePath("")
                    .displayName("Form1")
                    .lastStatusChangeDate(1487782554846L)
                    .formId("formId1")
                    .formVersion("2019103101")
                    .deletedDate(1487782554846L)
                    .geometryType("Point")
                    .geometry("{\"type\":\"Point\",\"coordinates\":[125.6, 10.0]}")
                    .status(Instance.STATUS_SUBMITTED).build(),

            new Instance.Builder()
                    .instanceFilePath("")
                    .displayName("Form2")
                    .lastStatusChangeDate(1488782558743L)
                    .formId("formId1")
                    .formVersion("2019103101")
                    .geometryType("Point")
                    .geometry("{\"type\":\"Point\",\"coordinates\":[125.6, 10.1]}")
                    .canEditWhenComplete(true)
                    .status(Instance.STATUS_COMPLETE).build(),

            new Instance.Builder()
                    .instanceFilePath("")
                    .displayName("Form3")
                    .lastStatusChangeDate(1484582553254L)
                    .formId("formId1")
                    .formVersion("2019103102")
                    .geometryType("Point")
                    .geometry("{\"type\":\"Point\",\"coordinates\":[126.6, 10.1]}")
                    .status(Instance.STATUS_INCOMPLETE).build(),

            new Instance.Builder()
                    .instanceFilePath("")
                    .displayName("Form4")
                    .lastStatusChangeDate(1488582557456L)
                    .formId("formId1")
                    .formVersion("2019103101")
                    .status(Instance.STATUS_COMPLETE).build(),

            new Instance.Builder()
                    .instanceFilePath("")
                    .displayName("Form5")
                    .lastStatusChangeDate(1483582557438L)
                    .formId("formId1")
                    .formVersion("2019103106")
                    .geometryType("Point")
                    .geometry("{\"type\":\"Point\",\"coordinates\":[125.6, 10.3]}")
                    .canEditWhenComplete(true)
                    .status(Instance.STATUS_SUBMISSION_FAILED).build(),

            new Instance.Builder()
                    .instanceFilePath("")
                    .displayName("Form6")
                    .lastStatusChangeDate(1482282559618L)
                    .formId("formId1")
                    .formVersion("2019103101")
                    .geometryType("Point")
                    .geometry("{\"type\":\"Point\",\"coordinates\":[125.7, 10.3]}")
                    .canEditWhenComplete(true)
                    .status(Instance.STATUS_SUBMITTED).build(),

            new Instance.Builder()
                    .instanceFilePath("")
                    .displayName("Form7")
                    .lastStatusChangeDate(1484782559836L)
                    .formId("formId1")
                    .formVersion("2019103101")
                    .geometryType("Point")
                    .geometry("{\"type\":\"Point\",\"coordinates\":[125.6, 10.4]}")
                    .canEditWhenComplete(false)
                    .status(Instance.STATUS_SUBMISSION_FAILED).build(),

            new Instance.Builder()
                    .instanceFilePath("")
                    .displayName("Form8")
                    .lastStatusChangeDate(1487982552254L)
                    .formId("formId2")
                    .formVersion("2019103101")
                    .geometryType("Point")
                    .geometry("Crazy stuff")
                    .status(Instance.STATUS_COMPLETE).build(),

            new Instance.Builder()
                    .instanceFilePath("")
                    .displayName("Form9")
                    .lastStatusChangeDate(1484682557369L)
                    .formId("formId2")
                    .formVersion("2019103101")
                    .geometryType("Crazy stuff")
                    .geometry("{\"type\":\"Point\",\"coordinates\":[125.6, 10.4]}")
                    .status(Instance.STATUS_COMPLETE).build(),
    };
}

File: collect_app/src/test/java/org/odk/collect/android/activities/FormMapActivityTest.kt
package org.odk.collect.android.activities

import android.app.Application
import android.content.Context
import android.content.Intent
import androidx.lifecycle.ViewModelProvider
import androidx.test.core.app.ActivityScenario
import androidx.test.core.app.ApplicationProvider.getApplicationContext
import androidx.test.ext.junit.runners.AndroidJUnit4
import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.Matchers.equalTo
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.kotlin.doReturn
import org.mockito.kotlin.mock
import org.odk.collect.android.R
import org.odk.collect.android.injection.config.AppDependencyModule
import org.odk.collect.android.storage.StoragePathProvider
import org.odk.collect.android.support.CollectHelpers
import org.odk.collect.android.utilities.FormsRepositoryProvider
import org.odk.collect.android.utilities.InstancesRepositoryProvider
import org.odk.collect.androidtest.ActivityScenarioLauncherRule
import org.odk.collect.forms.Form
import org.odk.collect.forms.instances.Instance
import org.odk.collect.formstest.FormUtils.buildForm
import org.odk.collect.formstest.InMemFormsRepository
import org.odk.collect.formstest.InMemInstancesRepository
import org.odk.collect.formstest.InstanceUtils.buildInstance
import org.odk.collect.geo.MappableSelectItem
import org.odk.collect.geo.SelectionMapViewModel
import org.odk.collect.shared.TempFiles.createTempDir
import java.text.SimpleDateFormat
import java.util.Locale

@RunWith(AndroidJUnit4::class)
class FormMapActivityTest {

    private val formsRepository = InMemFormsRepository()
    private val instancesRepository = InMemInstancesRepository()
    private val application = getApplicationContext<Application>()

    @get:Rule
    val launcherRule = ActivityScenarioLauncherRule()

    @Before
    fun setup() {
        CollectHelpers.setupDemoProject()

        val formsRepositoryProvider = mock<FormsRepositoryProvider> {
            on { get() } doReturn formsRepository
        }

        val instancesRepositoryProvider = mock<InstancesRepositoryProvider> {
            on { get() } doReturn instancesRepository
        }

        CollectHelpers.overrideAppDependencyModule(object : AppDependencyModule() {
            override fun providesFormsRepositoryProvider(application: Application?): FormsRepositoryProvider {
                return formsRepositoryProvider
            }

            override fun providesInstancesRepositoryProvider(
                context: Context?,
                storagePathProvider: StoragePathProvider?
            ): InstancesRepositoryProvider {
                return instancesRepositoryProvider
            }
        })
    }

    @Test
    fun `sets item count based on form instances`() {
        val form = formsRepository.save(
            buildForm("id", "version", createTempDir().absolutePath)
                .build()
        )
        instancesRepository.save(
            buildInstance(form.formId, form.version, createTempDir().absolutePath)
                .build()
        )
        instancesRepository.save(
            buildInstance(form.formId, form.version, createTempDir().absolutePath)
                .build()
        )

        val scenario = launchActivity(form)
        scenario.onActivity {
            val viewModel = ViewModelProvider(it)[SelectionMapViewModel::class.java]
            assertThat(viewModel.getItemCount().value, equalTo(2))
        }
    }

    @Test
    fun `sets items based on form instances with point geometry`() {
        val form = formsRepository.save(
            buildForm("id", "version", createTempDir().absolutePath)
                .build()
        )

        instancesRepository.save(
            buildInstance(form.formId, form.version, createTempDir().absolutePath)
                .geometry("")
                .geometryType("Something else")
                .build()
        )
        instancesRepository.save(
            buildInstance(form.formId, form.version, createTempDir().absolutePath)
                .build()
        )

        val instanceWithPoint = instancesRepository.save(
            buildInstance(form.formId, form.version, createTempDir().absolutePath)
                .geometry("{ \"coordinates\": [1.0, 2.0] }")
                .geometryType("Point")
                .canEditWhenComplete(true)
                .build()
        )

        val scenario = launchActivity(form)
        scenario.onActivity {
            val viewModel = ViewModelProvider(it)[SelectionMapViewModel::class.java]
            assertThat(viewModel.getMappableItems().value.size, equalTo(1))

            val expectedItem = MappableSelectItem(
                instanceWithPoint.dbId,
                2.0,
                1.0,
                R.drawable.ic_room_form_state_incomplete_24dp,
                R.drawable.ic_room_form_state_incomplete_48dp,
                instanceWithPoint.displayName,
                MappableSelectItem.IconifiedText(
                    R.drawable.form_state_saved,
                    formatDate(
                        R.string.saved_on_date_at_time,
                        instanceWithPoint.lastStatusChangeDate
                    )
                ),
                info = null,
                action = MappableSelectItem.IconifiedText(
                    R.drawable.ic_edit,
                    application.getString(R.string.review_data)
                )
            )
            assertThat(viewModel.getMappableItems().value[0], equalTo(expectedItem))
        }
    }

    @Test // Replace with a unit test
    fun `finalized instances with geometry have edit action and no info`() {
        val form = formsRepository.save(
            buildForm("id", "version", createTempDir().absolutePath)
                .build()
        )
        val instance = instancesRepository.save(
            buildInstance(form.formId, form.version, createTempDir().absolutePath)
                .geometry("{ \"coordinates\": [1.0, 2.0] }")
                .geometryType("Point")
                .canEditWhenComplete(true)
                .status(Instance.STATUS_COMPLETE)
                .build()
        )

        val scenario = launchActivity(form)
        scenario.onActivity {
            val viewModel = ViewModelProvider(it)[SelectionMapViewModel::class.java]
            val expectedItem = MappableSelectItem(
                instance.dbId,
                2.0,
                1.0,
                R.drawable.ic_room_form_state_complete_24dp,
                R.drawable.ic_room_form_state_complete_48dp,
                instance.displayName,
                MappableSelectItem.IconifiedText(
                    R.drawable.form_state_finalized,
                    formatDate(
                        R.string.finalized_on_date_at_time,
                        instance.lastStatusChangeDate
                    )
                ),
                info = null,
                action = MappableSelectItem.IconifiedText(
                    R.drawable.ic_edit,
                    application.getString(R.string.review_data)
                )
            )
            assertThat(viewModel.getMappableItems().value[0], equalTo(expectedItem))
        }
    }

    @Test // Replace with a unit test
    fun `deleted instances with geometry have info and no action`() {
        val form = formsRepository.save(
            buildForm("id", "version", createTempDir().absolutePath)
                .build()
        )
        val instance = instancesRepository.save(
            buildInstance(form.formId, form.version, createTempDir().absolutePath)
                .geometry("{ \"coordinates\": [1.0, 2.0] }")
                .geometryType("Point")
                .canEditWhenComplete(true)
                .deletedDate(123L)
                .build()
        )

        val scenario = launchActivity(form)
        scenario.onActivity {
            val viewModel = ViewModelProvider(it)[SelectionMapViewModel::class.java]
            val expectedItem = MappableSelectItem(
                instance.dbId,
                2.0,
                1.0,
                R.drawable.ic_room_form_state_incomplete_24dp,
                R.drawable.ic_room_form_state_incomplete_48dp,
                instance.displayName,
                MappableSelectItem.IconifiedText(
                    R.drawable.form_state_saved,
                    formatDate(
                        R.string.saved_on_date_at_time,
                        instance.lastStatusChangeDate
                    )
                ),
                info = formatDate(R.string.deleted_on_date_at_time, 123L),
                action = null
            )
            assertThat(viewModel.getMappableItems().value[0], equalTo(expectedItem))
        }
    }

    @Test // Replace with a unit test
    fun `submitted instances with geometry have view action and no info`() {
        val form = formsRepository.save(
            buildForm("id", "version", createTempDir().absolutePath)
                .build()
        )
        val instance = instancesRepository.save(
            buildInstance(form.formId, form.version, createTempDir().absolutePath)
                .geometry("{ \"coordinates\": [1.0, 2.0] }")
                .geometryType("Point")
                .canEditWhenComplete(true)
                .status(Instance.STATUS_SUBMITTED)
                .build()
        )

        val scenario = launchActivity(form)
        scenario.onActivity {
            val viewModel = ViewModelProvider(it)[SelectionMapViewModel::class.java]
            val expectedItem = MappableSelectItem(
                instance.dbId,
                2.0,
                1.0,
                R.drawable.ic_room_form_state_submitted_24dp,
                R.drawable.ic_room_form_state_submitted_48dp,
                instance.displayName,
                MappableSelectItem.IconifiedText(
                    R.drawable.form_state_submited,
                    formatDate(
                        R.string.sent_on_date_at_time,
                        instance.lastStatusChangeDate
                    )
                ),
                info = null,
                action = MappableSelectItem.IconifiedText(
                    R.drawable.ic_visibility,
                    application.getString(R.string.view_data)
                )
            )
            assertThat(viewModel.getMappableItems().value[0], equalTo(expectedItem))
        }
    }

    @Test // Replace with a unit test
    fun `instances that failed to submit with geometry have view action and no info`() {
        val form = formsRepository.save(
            buildForm("id", "version", createTempDir().absolutePath)
                .build()
        )
        val instance = instancesRepository.save(
            buildInstance(form.formId, form.version, createTempDir().absolutePath)
                .geometry("{ \"coordinates\": [1.0, 2.0] }")
                .geometryType("Point")
                .canEditWhenComplete(true)
                .status(Instance.STATUS_SUBMISSION_FAILED)
                .build()
        )

        val scenario = launchActivity(form)
        scenario.onActivity {
            val viewModel = ViewModelProvider(it)[SelectionMapViewModel::class.java]
            val expectedItem = MappableSelectItem(
                instance.dbId,
                2.0,
                1.0,
                R.drawable.ic_room_form_state_submission_failed_24dp,
                R.drawable.ic_room_form_state_submission_failed_48dp,
                instance.displayName,
                MappableSelectItem.IconifiedText(
                    R.drawable.form_state_submission_failed,
                    formatDate(
                        R.string.sending_failed_on_date_at_time,
                        instance.lastStatusChangeDate
                    )
                ),
                info = null,
                action = MappableSelectItem.IconifiedText(
                    R.drawable.ic_visibility,
                    application.getString(R.string.view_data)
                )
            )
            assertThat(viewModel.getMappableItems().value[0], equalTo(expectedItem))
        }
    }

    @Test // Replace with a unit test
    fun `instances that cannot be edited after completion with geometry have info and no action`() {
        val form = formsRepository.save(
            buildForm("id", "version", createTempDir().absolutePath)
                .build()
        )
        instancesRepository.save(
            buildInstance(form.formId, form.version, createTempDir().absolutePath)
                .geometry("{ \"coordinates\": [1.0, 2.0] }")
                .geometryType("Point")
                .canEditWhenComplete(false)
                .status(Instance.STATUS_COMPLETE)
                .build()
        )
        instancesRepository.save(
            buildInstance(form.formId, form.version, createTempDir().absolutePath)
                .geometry("{ \"coordinates\": [1.0, 2.0] }")
                .geometryType("Point")
                .canEditWhenComplete(false)
                .status(Instance.STATUS_SUBMITTED)
                .build()
        )
        instancesRepository.save(
            buildInstance(form.formId, form.version, createTempDir().absolutePath)
                .geometry("{ \"coordinates\": [1.0, 2.0] }")
                .geometryType("Point")
                .canEditWhenComplete(false)
                .status(Instance.STATUS_SUBMISSION_FAILED)
                .build()
        )

        val scenario = launchActivity(form)
        scenario.onActivity {
            val viewModel = ViewModelProvider(it)[SelectionMapViewModel::class.java]
            val items = viewModel.getMappableItems().value

            assertThat(items[0].action, equalTo(null))
            assertThat(
                items[0].info,
                equalTo(application.getString(R.string.cannot_edit_completed_form))
            )

            assertThat(items[1].action, equalTo(null))
            assertThat(
                items[1].info,
                equalTo(application.getString(R.string.cannot_edit_completed_form))
            )

            assertThat(items[2].action, equalTo(null))
            assertThat(
                items[2].info,
                equalTo(application.getString(R.string.cannot_edit_completed_form))
            )
        }
    }

    private fun launchActivity(form: Form): ActivityScenario<FormMapActivity> {
        val intent = Intent(application, FormMapActivity::class.java).also {
            it.putExtra(FormMapActivity.EXTRA_FORM_ID, form.dbId)
        }
        val scenario = launcherRule.launch<FormMapActivity>(intent)
        return scenario
    }

    private fun formatDate(string: Int, date: Long): String {
        return SimpleDateFormat(
            application.getString(string),
            Locale.getDefault()
        ).format(date)
    }
}
